### ISSUE TITLE
question: Email setup

### ISSUE THREAD

**Issue by samcro1967** [greenpau/caddy-security#50]
Trying to get email working for user registration.  I do not see any errors in the logs, but get email error when a user tries to register.  Below is the relevant config I am using.  Guessing I don't quite have the config correct, but not sure what I am missing.

```
	security {
		credentials user@mydomain.net {
			username user@mydomain.net
			password <app_passowrd>
		}
		messaging email provider gmail {
			address smtp.gmail.com:587
			protocol smtp
			credentials user@mydomain.net
			sender user@mydomain.net "My Auth Portal"
		}
		authentication portal myportal {
			enable source ip tracking
			cookie lifetime 86400
			crypto default token lifetime 3600
			crypto key sign-verify <redacted>
			cookie domain mydomain.net
			backend local /config/caddy/users.json local
			transform user {
				match roles registered
				require mfa
			}
			registration {
				dropbox /config/caddy/registrations.json
				title "User Registration"
				code "Test"
				require domain mx
				email provider gmail
			}
...
```

**Comment by samcro1967** [comment]
Tried it with stmps also, with no luck.  Added the line below to my Dockerfile.  Still not seeing anything in the container logs related to the user registration email failure.  Not noticing anything different in the logs compiled with the latest commit.

**Dockerfile**
```
RUN xcaddy build \
    --with github.com/lucaslorentz/caddy-docker-proxy/plugin/v2 \
    --with github.com/greenpau/caddy-security \
    --with github.com/greenpau/caddy-trace \
    -with github.com/porech/caddy-maxmind-geolocation \
    --with github.com/caddy-dns/cloudflare \
    --with github.com/kirsch33/realip \
    --with github.com/greenpau/go-authcrunch@c42d9cae65c8bab5b047a80ece6b9d6b3b083ea6
```

**Caddyfile global block**
```
	http_port	8069
	https_port	443
	email user@mydoamin.net
	debug
	log {
		output file /data/caddy.log
                level warn
		format console
        }
```

```
{"level":"info","ts":1644543830.1393936,"logger":"docker-proxy","msg":"Running caddy proxy server"}
{"level":"info","ts":1644543830.1402934,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["localhost:2019","[::1]:2019","127.0.0.1:2019"]}
{"level":"info","ts":1644543830.1405373,"msg":"autosaved config (load with --resume flag)","file":"/config/caddy/autosave.json"}
{"level":"info","ts":1644543830.1405458,"logger":"docker-proxy","msg":"Running caddy proxy controller"}
{"level":"info","ts":1644543830.1411324,"logger":"docker-proxy","msg":"Start","CaddyfilePath":"/data/CaddyFile","LabelPrefix":"caddy","PollingInterval":30,"ProcessCaddyfile":true,"ProxyServiceTasks":true,"IngressNetworks":"[]"}
{"level":"info","ts":1644543830.1412792,"logger":"docker-proxy","msg":"Caddy ContainerID","ID":"f3cc0b58aed2cf5ea660cfef700480f57020512c4b90a3f9a280eb77cec6047b"}
{"level":"info","ts":1644543830.144751,"logger":"docker-proxy","msg":"Connecting to docker events"}
{"level":"info","ts":1644543830.192845,"logger":"docker-proxy","msg":"IngressNetworksMap","ingres":"map[12d99874b921591e07c8c2c0c50f6fb013984cd9045b8b0dda296fea3ce8cc18:true]"}
{"level":"info","ts":1644543830.2060375,"logger":"docker-proxy","msg":"Swarm is available","new":false}
{"level":"info","ts":1644543830.2063546,"logger":"docker-proxy","msg":"Skipping swarm config caddyfiles because swarm is not available"}
{"level":"info","ts":1644543830.2387893,"logger":"docker-proxy","msg":"Skipping swarm services because swarm is not available"}

Caddyfile in console and JSON format.

{"level":"info","ts":1644543830.2778704,"logger":"docker-proxy","msg":"Sending configuration to","server":"localhost"}
{"level":"info","ts":1644543830.2790034,"logger":"admin.api","msg":"received request","method":"POST","host":"localhost:2019","uri":"/load","remote_addr":"127.0.0.1:36798","headers":{"Accept-Encoding":["gzip"],"Content-Length":["23538"],"Content-Type":["application/json"],"User-Agent":["Go-http-client/1.1"]}}
{"level":"info","ts":1644543830.28317,"msg":"redirected default logger","from":"stderr","to":"/data/caddy.log"}
{"level":"info","ts":1644543830.4065075,"logger":"docker-proxy","msg":"Successfully configured","server":"localhost"}
```

**Comment by greenpau** [comment]
> Caddyfile in console and JSON format.

@samcro1967, do you see the email provider configuration in the JSON?

**Comment by greenpau** [comment]
@samcro1967 , please take a screenshot of the error the user is receiving. Generally speaking, the error should show up in the logs.

**Comment by samcro1967** [comment]
A grep of the logs for error returns no results.

![image](https://user-images.githubusercontent.com/2319445/153532476-1418f4a7-d9f5-4b5a-a9f5-764b3f5c3955.png)

**Comment by greenpau** [comment]
@samcro1967 , the code that triggers the errors follows:

```
                p.logger.Warn(
                    "Failed to send notification",
                    zap.String("session_id", rr.Upstream.SessionID),
                    zap.String("request_id", rr.ID),
                    zap.String("registration_id", registrationID),
                    zap.String("registration_type", "registration_confirmation"),
                    zap.Error(err),
                )
                p.registrations.Delete(registrationID)
                message = "Internal registration messaging error"
```

There should be an error that says `Failed to send notification`.

**Comment by greenpau** [comment]
@samcro1967 , please try upgrading to [v1.0.14](https://github.com/greenpau/caddy-security/releases/tag/v1.0.14). 

You would need to provide the `admin email` directive:

```
			registration {
				dropbox /config/caddy/registrations.json
				title "User Registration"
				code "Test"
				require domain mx
				email provider gmail
				admin email foobar@foo.bar
			}
```

**Comment by samcro1967** [comment]
Having issues getting v1.0.14.

container build log shows
```
2022/02/11 12:55:32 [INFO] exec (timeout=0s): /usr/local/go/bin/go get -d -v github.com/greenpau/caddy-security
go: downloading github.com/greenpau/caddy-security v1.0.14
```

docker exec -it caddy /bin/sh -c "caddy list-modules -versions | grep security" shows "security v1.0.13".

Also tried specifying "--with github.com/greenpau/caddy-security@v1.0.14" in the Dockerfile and see the same results.

Removed the container and all images and rebuilt to make sure there is only one image and one container running that image.  I don't think I have any wires crossed anywhere....

**Comment by greenpau** [comment]
@samcro1967 , please paste Dockerfile “—with” statements.

**Comment by samcro1967** [comment]
Complete Dockerfile

```
ARG CADDY_VERSION=2.4.6

FROM caddy:${CADDY_VERSION}-builder AS builder

RUN xcaddy build \
    --with github.com/lucaslorentz/caddy-docker-proxy/plugin/v2 \
    --with github.com/greenpau/caddy-security@v1.0.14 \
    --with github.com/greenpau/caddy-trace \
    --with github.com/porech/caddy-maxmind-geolocation \
    --with github.com/caddy-dns/cloudflare \
    --with github.com/kirsch33/realip \
    --with github.com/greenpau/go-authcrunch@c42d9cae65c8bab5b047a80ece6b9d6b3b083ea6

FROM caddy:${CADDY_VERSION}-alpine

# Timezone
ENV TIMEZONE America/Chicago

# Install timezone data and setup timezone
RUN apk --update add --no-cache --virtual .tz-deps tzdata \
    && cp /usr/share/zoneinfo/${TIMEZONE} /etc/localtime \
    && echo ${TIMEZONE} > /etc/timezone \
    && apk del .tz-deps

COPY --from=builder /usr/bin/caddy /usr/bin/caddy

CMD ["caddy", "docker-proxy"]
```

**Comment by greenpau** [comment]
@samcro1967 , please remove

--with github.com/greenpau/go-authcrunch@c42d9cae65c8bab5b047a80ece6b9d6b3b083ea6


It is no longer needed

**Comment by samcro1967** [comment]
Here is what I am seeing in the access log.  I didn't redact anything.  Nothing looks senstivie.
```
1.6445883683845236e+09  warn    security        Failed to send notification     {"session_id": "dqH7qFHUUcCQApZssCNlhzgLXJyKHbf44vb6CsYILs2b", "request_id": "ebe7bf36-7a0e-4b79-b8fe-3106394a3915", "registration_id": "y37lU6CrDI1p79IWfwXoz4HA9WisxmYI7XZn8EG5F62l5m4uEtz85bhee5ue9ctKG4anCCKg43lMmvkyw7V4qELYZi6Y4O4", "registration_type": "registration_confirmation", "error": "notification request via \"gmail\" email provider failed: tls: first record does not look like a TLS handshake"}
```

**Comment by greenpau** [comment]
> "error": "notification request via \"gmail\" email provider failed: tls: first record does not look like a TLS handshake"}

@samcro1967 , this looks like connectivity issue to Gmail. I will try adding extra debugging over the upcoming weekend.

**Comment by greenpau** [comment]
@samcro1967 , do you have proxies in your environment? https://stackoverflow.com/questions/61907287/docker-first-record-does-not-look-like-a-tls-handshake-error-when-pull-image

**Comment by samcro1967** [comment]
There is no proxy between the caddy container and the Internet.  I am able to ping ping smtp.gmail.com from the container.

/srv # ping smtp.gmail.com
PING smtp.gmail.com (74.125.129.108): 56 data bytes
64 bytes from 74.125.129.108: seq=0 ttl=101 time=30.878 ms

**Comment by greenpau** [comment]
@samcro1967 , another suggestion “think, you should be using port 465 instead of 587.”

**Comment by samcro1967** [comment]
I am using 587 on a handful other other containers successfully for messaging.  Seeing "Username and Password not accepted" on port 465.  I do have MFA enabled for gmail so I am using an app generated password.  Uptime-Kuma works on port 587, but not 465.  Guessing app generated passwords do not work for auth on port 465.

```
1.6445895953141277e+09  warn    security        Failed to send notification     {"session_id": "ZtTf1eLzg0xiA5FxWs5pLLukwfSfItUSeAaw6er7c7Zt", "request_id": "ee35a9ff-4a71-417d-a487-0a4fa43785b7", "registration_id": "uOoHaTmfqYXJOlGvLUmR127dhDGlZ6OJFvjL4RZt5KbVGVuELHwqBJo2mhf0VtMDA", "registration_type": "registration_confirmation", "error": "notification request via \"gmail\" email provider failed: Username and Password not accepted. Learn more at\n https://support.google.com/mail/?p=BadCredentials l12sm13850753iow.48 - gsmtp"}
```

**Comment by greenpau** [comment]
> so I am using an app generated password. Uptime-Kuma works on port 587, but not 465. Guessing app generated passwords do not work for auth on port 465.

@samcro1967 , please describe the process you have used to used to create this custom password. I could try replicating it with my Gmail.

**Comment by samcro1967** [comment]
Here are the instructions on how to setup an app password.
https://support.google.com/mail/answer/185833?hl=en

**Comment by greenpau** [comment]
> Here are the instructions on how to setup an app password. https://support.google.com/mail/answer/185833?hl=en

@samcro1967 , thank you! 👍  I will try it.

**Comment by greenpau** [comment]
@samcro1967 , I generated the app password and the following configuration worked for me.

```
        credentials myuser@gmail.com {
            username {env.GMAIL_USER}
            password {env.GMAIL_SECRET}
        }

        messaging email provider gmail-smtp-server {
            address smtp.gmail.com:587
            protocol smtp
            sender myuser@gmail.com
            credentials myuser@gmail.com
        }
```

**Comment by samcro1967** [comment]
I was able to get it to work following your last config above, but I had to generate a new app password.  It would not work with an existing app password.  I am using the same app password for a handful of existing apps without issue.  Not sure why, but it is working now.  Works nicely and is very clean.

Is the step to move the user from registrations.json to users.json a manual step?  I don't see a place in auth/settings to approve or decline the registration.  Or is there a different path for the management interface?
